### ISSUE TITLE
 [cordova-debug-adapter] Error: Unable to find pid

### ISSUE THREAD

**Issue by sahildaga95** [Microsoft/vscode-cordova#186]
Hello,
I am developing my app using ionic and i wanted to debug my application on Mobile device not emulator.It works perfectly in Windows but do not launch on ubuntu or macintosh.
I also have usb debugging enabled on device,as i said works perfect in Windows.
This is my ionic info
Cordova CLI: 6.3.1
Ionic CLI Version: 2.0.0-beta.32
Ionic App Lib Version: 2.0.0-beta.18
OS: Distributor ID: Ubuntu Description: Ubuntu 16.04 LTS
Node Version: v6.4.0

This is my launch.json,

```
{
"version": "0.2.0",
"configurations": [
    {
        "name": "Run Android on device",
        "type": "cordova",
        "request": "launch",
        "platform": "android",
        "target": "device",
        "port": 9222,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}",
        "ionicLiveReload": false
    },
    {
        "name": "Run iOS on device",
        "type": "cordova",
        "request": "launch",
        "platform": "ios",
        "target": "device",
        "port": 9220,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}",
        "ionicLiveReload": false
    },
    {
        "name": "Attach to running android on device",
        "type": "cordova",
        "request": "attach",
        "platform": "android",
        "target": "device",
        "port": 9222,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}"
    },
    {
        "name": "Attach to running iOS on device",
        "type": "cordova",
        "request": "attach",
        "platform": "ios",
        "target": "device",
        "port": 9220,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}"
    },
    {
        "name": "Run Android on emulator",
        "type": "cordova",
        "request": "launch",
        "platform": "android",
        "target": "emulator",
        "port": 9222,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}",
        "ionicLiveReload": false
    },
    {
        "name": "Run iOS on simulator",
        "type": "cordova",
        "request": "launch",
        "platform": "ios",
        "target": "emulator",
        "port": 9220,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}",
        "ionicLiveReload": false
    },
    {
        "name": "Attach to running android on emulator",
        "type": "cordova",
        "request": "attach",
        "platform": "android",
        "target": "emulator",
        "port": 9222,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}"
    },
    {
        "name": "Attach to running iOS on simulator",
        "type": "cordova",
        "request": "attach",
        "platform": "ios",
        "target": "emulator",
        "port": 9220,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}"
    },
    {
        "name": "Serve to the browser (ionic serve)",
        "type": "cordova",
        "request": "launch",
        "platform": "serve",
        "cwd": "${workspaceRoot}",
        "devServerAddress": "localhost",
        "sourceMaps": true,
        "ionicLiveReload": true
    },
    {
        "name": "Simulate Android in browser",
        "type": "cordova",
        "request": "launch",
        "platform": "android",
        "target": "chrome",
        "simulatePort": 8000,
        "livereload": true,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}"
    },
    {
        "name": "Simulate iOS in browser",
        "type": "cordova",
        "request": "launch",
        "platform": "ios",
        "target": "chrome",
        "simulatePort": 8000,
        "livereload": true,
        "sourceMaps": true,
        "cwd": "${workspaceRoot}"
    }
]
```

}

**Comment by MSLaguana** [comment]
@sahildaga95: When you try to debug your app on a device using Ubuntu or mac, what errors do you see? Are you just trying to debug an android app, or are you also trying to do iOS from mac? What happens if you run `cordova run android` on those machines?

**Comment by sahildaga95** [comment]
Actually i am making an ionic 2 application so i use the command ionic run android , it creates a build , install the apk in the device and run the app on mobile but i cant use visual studio to debug it. Since i cant apply breakpoints in my running code. It will be only possible if i  use "Run Android on device" from debug window. I think there might be some changes in launch.json since " Run android on device" works perfectly on windows .

**Comment by MSLaguana** [comment]
What error do you get when you try to `Run android on device` in VS code?

**Comment by sahildaga95** [comment]
And the error i see is "[cordova-debug-adapter] Error: Unable to find pid" it detects the device but is unable to launch the app on device.

**Comment by MSLaguana** [comment]
Do you have `adb` on your PATH? If you try to run your app using VSCode, does the app start successfully, but then not debug? We try to find the process ID of the application on the device once we start it so we can hook up the debugging channel, and that error indicates we cannot find the process.

**Comment by MSLaguana** [comment]
One other thing I'd like to check; do you have a `platforms/android/AndroidManifest.xml` which has a `package` attribute on the root element?

**Comment by sahildaga95** [comment]
No the app doesn't start it shows the error.
 This is my AndroidMainfest.xml

            <?xml version='1.0' encoding='utf-8'?>
            <manifest android:hardwareAccelerated="true" android:versionCode="1" android:versionName="0.0.1" package="com.ionicframework.ionic2formvalidation544845" xmlns:android="http://schemas.android.com/apk/res/android">
                <supports-screens android:anyDensity="true" android:largeScreens="true" android:normalScreens="true" android:resizeable="true" android:smallScreens="true" android:xlargeScreens="true" />
                <uses-permission android:name="android.permission.INTERNET" />
                <application android:hardwareAccelerated="true" android:icon="@drawable/icon" android:label="@string/app_name" android:supportsRtl="true">
                    <activity android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale" android:label="@string/activity_name" android:launchMode="singleTop" android:name="MainActivity" android:theme="@android:style/Theme.DeviceDefault.NoActionBar" android:windowSoftInputMode="adjustResize">
                        <intent-filter android:label="@string/launcher_name">
                            <action android:name="android.intent.action.MAIN" />
                            <category android:name="android.intent.category.LAUNCHER" />
                        </intent-filter>
                    </activity>
                </application>
                <uses-sdk android:minSdkVersion="16" android:targetSdkVersion="23" />
            </manifest>

**Comment by MSLaguana** [comment]
Do you have version 1.2.2 of our extension installed on all these machines? We should just be invoking `ionic run android --device --verbose` to start the app. When you run that yourself, it works fine, but when you run vscode it does not?

How is ionic placed on your PATH? Is it done as part of a `.bashrc` script or something, or is it via a symlink in `/usr/bin/` or similar which should work all the time?

**Comment by sahildaga95** [comment]
Yes i have same version on all these machines.
Yes when i run using ionic run android it works and does not run through vscode.
Ionic is placed at path /home/user/node_modules/ionic 
I did not setup anything additional in .bashrc or anything else.

But it started launching automatically on one of my friends linux machine , which displayed the same issue like me and he didnt do any changes to his system as well.
What could be the reason for that ??

**Comment by MSLaguana** [comment]
You say that ionic is in `/home/user/node_modules/ionic`, how is that placed on your PATH? Have you tried running vscode from the command line (should be a `code` alias to start it I believe), and see if that works?

**Comment by sahildaga95** [comment]
I followed the commands given at ionic 2 website and used " sudo npm install -g ionic@beta  " to install ionic.

**Comment by MSLaguana** [comment]
Could you please make sure that vscode is closed (on mac you need to go via the taskbar entry, not just closing the windows), navigate to your project on the command line and try `code .` to open your project in vscode, and see if the issue still happens there?

**Comment by jicongw** [comment]
Hey @sahildaga95 , I verified this scenario on my Mac machine and it seems to be working. Is this still happening for you?

**Comment by sahildaga95** [comment]
@MSLaguana : We are still not able to resolve this issue. I followed below steps for installation and configuration on Ubuntu-14.04 and 16.04 with normal user 'lg'.

1) Download VSCODE from https://go.microsoft.com/fwlink/?LinkID=760868 and istalled it.
2) I followed URL for intallation of Nodejs 6.4.0 " http://sysads.co.uk/2016/08/how-to-install-node-js-6-4-0-on-ubuntu-16-04/ "
3) With normal user 'lg' i executed below commands $ npm install -g cordova , $ npm install -g ionic@beta
4) Downloaded Android SDK and placed into users home directory also set ANDROID_HOME in .profile . https://developer.android.com/studio/releases/sdk-tools.html#notes
5) Then i successfully launched vscode using $ code .
6) I installed cordova tools  and chrome debugger plugins
7)Launched my ionic app using "Run Android on device". (Works perfectly when i launched app using ionic serve on Browser)
8)When tried to launch on device following error message displayed on debugging console :-

        Launching for android (This may take a while)...
        App successfully launched
        Attaching to android
        Unable to find pid of cordova app

**Comment by amnonjw** [comment]
I run VSCode, updated extensions for debugging, and I'm developing a project with Ionic (ver 1).
I still, have issues with debugging on device and I often get the same output as shown here. My device is Nexus 5x (running Android Marshmallow).

I know that when it does work, VSCode installs a temp version app and runs it, then probably attaches to it. However, when this error occurs, I don't see any new app installed. It is as if the plugin builds the app and "installs" it somewhere else

**Comment by MSLaguana** [comment]
We install the app on the device by running `ionic run android --device --verbose`. The next time we fail to debug your app, could you please try manually running that command in your project?

**Comment by amnonjw** [comment]
I opened a post 
https://github.com/Microsoft/vscode-cordova/issues/98
with similar issue on an emulator

Anyways, since you asked for the output of said command
(it is actually very long, I paste here the end of it)

> :processDebugManifest
> :processDebugResources
> :generateDebugSources
> :compileDebugJavaWithJavac**_Note: Some input files use or override a deprecated AP
> I.
> Note: Recompile with -Xlint:deprecation for details.
> Note: Some input files use unchecked or unsafe operations.
> Note: Recompile with -Xlint:unchecked for details._**
> 
> :compileDebugNdk UP-TO-DATE
> :compileDebugSources
> :transformClassesWithDexForDebug
> :mergeDebugJniLibFolders UP-TO-DATE
> :transformNative_libsWithMergeJniLibsForDebug UP-TO-DATE
> :processDebugJavaRes UP-TO-DATE
> :transformResourcesWithMergeJavaResForDebug UP-TO-DATE
> :validateDebugSigning
> :packageDebug
> :zipalignDebug
> :assembleDebug
> :cdvBuildDebug

> BUILD SUCCESSFUL
> Total time: 22.298 secs
> Built the following apk(s):
>         c:/IonicProjects/SocialClub/platforms/android/build/outputs/apk/android-
> debug.apk
> Using apk: c:/IonicProjects/SocialClub/platforms/android/build/outputs/apk/andro
> id-debug.apk
> Installing apk c:/IonicProjects/SocialClub/platforms/android/build/outputs/apk/a
> ndroid-debug.apk on 00f43c04f83039e3...
> Running command "input keyevent 82" on 00f43c04f83039e3...
> Starting application "com.ionicframework.socialclub/.MainActivity" on 00f43c04f8
> 3039e3...
> Running command "am start -W -a android.intent.action.MAIN -ncom.ionicframework.
> socialclub/.MainActivity" on 00f43c04f83039e3...
> LAUNCH SUCCESS
> Executing "after_run"  hook for all plugins.


and after that the app was running on my device. I executed this command twice, just to make sure, and the second time the lines I marked in bold were not there. I will write in another post a few "tests" I ran just now to see what is working and what is not.

**Comment by amnonjw** [comment]
First things first:
I executed the command you suggested `ionic run android --device --verbose` from a cmd.exe window which I opened with admin privileges. When VSCode runs the debug feature, I see in my tasks-manager the process
`node "c:\Users\<UserName>\AppData\Roaming\npm\\node_modules\cordova\bin\cordova" run android --device --verbose`

now... after using the cmd command I returned to VSCode and 
* executed "run android on device". It worked
* uninstalled the "temp-app" from my device, executed "run android on device". It worked
* closed VSCode and the cmd window. uninstalled the "temp-app" again. Reopened VSCode, and tried "run android on device" - it did **NOT** worked.
* I closed VSCode and opened it with admin privileges. executed "run android on device" and it worked

Now I performed a small change in my code.
* I uninstalled the "temp-app". I tried "run android on device" - it did **NOT** worked.
* I closed VSCode and re-opened it with admin privileges. executed "run android on device" and it worked.
* Just for checking, I tried again but it did **NOT** worked and I received the following output
> Launching for android (This may take a while)...
> App successfully launched
> Attaching to android
> Forwarding debug port
> Attaching to app.
> 
> Got a response from the target app, but no target pages found
> error: listener not found
> error: listener not found
* I tried again - it worked.

Bottom line, the entire debug feature is not running smoothly, as if it needs the right conditions for it. Maybe when VSCode tries to run its command it executes several processes simultaneously (as I see in my tasks-manager) and one of them starts too early and the entire thing's is failing. Since I cannot see the full output log (only the four color lines VSCode shows me), I cannot know for sure.

p.s.
I tried putting a breakpoint inside `$ionicPlatform.ready(function() {` but it appears that when debugging on device (when it works) the breakpoint isn't triggered fast enough, and skipped. I can know that ionic platform is initiated in another way, but still it would have been nice to have.

**Comment by sahildaga95** [comment]
I am still facing the issue, Not always but when i do then its a complete blocker and i am unable to debug My App on mobile device. I have found a few ways to get ahead of it but sometimes it still won't work.These are few ways:-
1. Restart Mobile Device
2. Restart your computer
3.If there is a build already installed in the device then uninstall it and try again.
4 Run ionic serve in the background and launch the app through vscode plugin
5. Remove your Platform and Add it again
As i said earlier sometimes none of it works and after a day or two it starts working.

**Comment by gianpaj** [comment]
I had a similar issue and it was because I had error in my ionic2 app which made `ngc` fail.
```
[00:31:20]  Error: Error at /.../.tmp/pages/page2/page2.ngfactory.ts:393:169
[00:31:20]  Property 'Date' does not exist on type 'Page2'.
[00:31:20]  ngc failed
[00:31:20]  ionic-app-script task: "build"
[00:31:20]  Error: Error
```
This issue didn't show up in the VSCode Debug console

**Comment by nikeshwarpandey** [comment]
What are steps to create APK in VS code?

**Comment by amnonjw** [comment]
@nikeshwarpandey 
If you're asking this to know how to make a standalone apk of your app - I think that's a different topic. Anyways, you need to sign the apk in order to produce a public apk for devices, and VS Code doesn't let you do that, as far as I know. For cordova apps there are explanations online.

For debugging on devices, @MSLaguana wrote that the VS Code plugin uses `ionic run android --device --verbose` and that itself generates a self-sign-ish (I think) apk for debugging only. So as far as I know, VS Code doesn't generate apk by itself.

**Comment by mysterytony** [comment]
`01_jshint.js` would print something wrong with the code like missing a semi-colon, but vscode only display `unable to find pid` which is misleading.

**Comment by MSLaguana** [comment]
I believe that the issue here is we run `ionic run android` to launch the app on a device, and that command is always returning "successfully" even if it fails. We look at the output to try and detect certain failures, but there are many more cases that indicate something has gone wrong and the app has not successfully launched. The only way we find this out is when we try to attach to the app on the device, and instead we are unable to locate the running app (because it was not actually started by `ionic run android`).

**Comment by amnonjw** [comment]
And there is currently no additional way to get indication for errors?
When `ionic run android --device --verbose` is executed, it takes it time to deploy the app and running it. This means that some step are taken in a particular order, and the detailed output it gives - that shows indication for errors along the way, right? Do you monitor that kind of process other than waiting for a return value from the command and finding a PID?

**Comment by MSLaguana** [comment]
We do look at the output and try to detect errors, but only certain kinds of errors that we expect, while there can be essentially arbitrary errors that happen (especially with user scripts). Parsing textual output is always a risky business since we don't know exactly what an error might look like, and that's what has lead us here at the moment.

**Comment by sahildaga95** [comment]
So this is not an issue anymore I haven't encountered this error in a long time now. Thanks for the help.
